### PR TITLE
Only enable KTLS if it is explicitly configured

### LINKED ISSUE #13794: KTLS should not be on by default
At the moment KTLS support is not built by default. But if built it is on at runtime by default.

This might cause problems. For example If you are using the FIPS module you might reasonable expect all your crypto to be performed inside the module boundary. However, if you are using a KTLS enabled  build then that might not be the case if libssl decided it can offload things to the kernel.

A suggestion is to switch the KTLS defaults around - so that it is always built by default if possible, but it is disabled at runtime unless you explicitly ask for it.

### PR DESCRIPTION
**Pull request by mattcaswell**
It has always been the case that KTLS is not compiled by default. However
if it is compiled then it was automatically used unless specifically
configured not to. This is problematic because it avoids any crypto
implementations from providers. A user who configures all crypto to use
the FIPS provider may unexpectedly find that TLS related crypto is actually
being performed outside of the FIPS boundary.

Instead we change KTLS so that it is disabled by default.

We also swap to using a single "option" (i.e. SSL_OP_ENABLE_KTLS) rather
than two separate "modes", (i.e. SSL_MODE_NO_KTLS_RX and
SSL_MODE_NO_KTLS_TX).

Fixes #13794

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -20,6 +20,7 @@ OpenSSL 3.0
 
 ### Major changes between OpenSSL 1.1.1 and OpenSSL 3.0 [under development]
 
+  * Added suport for Kernal TLS (KTLS)
```
Comment (t8m) [inline NEWS.md:23]: typo: Kernal

File: doc/man3/SSL_CTX_set_options.pod [hunk 2]
```diff
@@ -279,6 +279,29 @@ responsible for cleansing all other buffers. Most notably, this
...
+have been compiled with support for it, and it must be supported by the
+negotiated ciphersuites and extensions. The specific ciphersuites and extensions
+that are supported may vary by platform and kernel version.
+
+The kernel TLS data-path implements the record layer, and the encryption
+algorithm. The kernel will utilize the best hardware
+available for encryption. Using the kernel data-path should reduce the memory
+footprint of OpenSSL because no buffering is required. Also, the throughput
+should improve because data copy is avoided when user data is encrypted into
+kernel memory instead of the usual encrypt than copy to kernel.
```
Comment (t8m) [inline doc/man3/SSL_CTX_set_options.pod:294]: Typo: than

File: doc/man3/SSL_CTX_set_options.pod [hunk 3]
```diff
@@ -279,6 +279,29 @@ responsible for cleansing all other buffers. Most notably, this
...
+
+Kernel TLS might not support all the features of OpenSSL. For instance,
+renegotiation, and setting the maximum fragment size is not possible as of
+Linux 4.20.
+
+Note that with kernel TLS enabled some cryptographic operations are performed
+by the kernel directly and not via any available OpenSSL Providers. This might
+be undesirable if, for example, the application requires all cryptographic
+operations to be performmed by the FIPS provider.
```
Comment (t8m) [inline doc/man3/SSL_CTX_set_options.pod:303]: Typo: performmed

**Comment by mattcaswell** [comment]
As part of this I updated various documentation including adding a CHANGES entry for the KTLS feature which was missing. I credited @borisPis-zz, @bsdjhb and @gallatin. Hopefully I got the attribution right for the most significant contributors to this.

File: ssl/ssl_conf.c [hunk 4]
```diff
@@ -386,7 +386,8 @@ static int cmd_Options(SSL_CONF_CTX *cctx, const char *value)
         SSL_FLAG_TBL("MiddleboxCompat", SSL_OP_ENABLE_MIDDLEBOX_COMPAT),
         SSL_FLAG_TBL_INV("AntiReplay", SSL_OP_NO_ANTI_REPLAY),
         SSL_FLAG_TBL_INV("ExtendedMasterSecret", SSL_OP_NO_EXTENDED_MASTER_SECRET),
-        SSL_FLAG_TBL_INV("CANames", SSL_OP_DISABLE_TLSEXT_CA_NAMES)
+        SSL_FLAG_TBL_INV("CANames", SSL_OP_DISABLE_TLSEXT_CA_NAMES),
+        SSL_FLAG_TBL("KTLS", SSL_OP_ENABLE_KTLS),
```
Comment (t8m) [inline ssl/ssl_conf.c:390]: Shouldn't the comma after last element break the ansi build? Hmm, dunno why it doesn't.

File: include/openssl/ssl.h.in [hunk 5]
```diff
@@ -330,6 +325,9 @@ typedef int (*SSL_async_callback_fn)(SSL *s, void *arg);
 /* Allow initial connection to servers that don't support RI */
 # define SSL_OP_LEGACY_SERVER_CONNECT                    0x00000004U
 
+/* Enable support for Kernel TLS */
+# define SSL_OP_ENABLE_KTLS                              0x00000008U
+
```
Comment (t8m) [inline include/openssl/ssl.h.in:330]: You did not delete the SSL_MODE_NO_KTLS_* defines.

**Review by t8m — CHANGES_REQUESTED** [review]

File: ssl/ssl_conf.c (hunk 4 above)
Comment (mattcaswell) [reply ssl/ssl_conf.c]: I removed the comma

**Comment by mattcaswell** [comment]
All feedback addressed. Please take another look.

**Review by t8m — APPROVED** [review]

**Review by bsdjhb — COMMENTED** [review]
I generally thinks this looks fine.  The sslapitest.c changes conflict quite a bit with the changes I have to refactor those in #13752.

I think for KTLS tests we might have to refactor how these tests work a bit more after this, e.g. we might need to run a specific test with KTLS enabled in the config (different config file perhaps?).  We would also want the tests to not report bogus failures if KTLS doesn't "take" for a given connection since the given cipher suite might not be supported by the drivers loaded at the time (e.g. in FreeBSD, TLS RX for MTE cipher suites only works if you are using a Chelsio NIC with TCP offload, whereas AEAD cipher suites for TLS 1.2 work out of the box for both RX and TX).  However, figuring out how to make the tests more sane is my only real wrinkle.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Comment by vvfedorenko** [comment]
@mattcaswell Could we rethink a bit more about single "option" (SSL_OP_ENABLE_KTLS)? It looks like current implementation of RX part in Linux kernel is not that great because of multiple additional copies done in crypto subsystem. Is it possible to have old SSL_MODE_NO_KTLS_{RX,TX} in coming versions? Users can benefit from choosing which part to offload and which part to do in userspace?

**Comment by mattcaswell** [comment]
> Could we rethink a bit more about single "option" (SSL_OP_ENABLE_KTLS)? It looks like current implementation of RX part in Linux kernel is not that great because of multiple additional copies done in crypto subsystem. Is it possible to have old SSL_MODE_NO_KTLS_{RX,TX} in coming versions?

Well, we can't change it now for 3.0 and we at least need to continue to support SSL_OP_ENABLE_KTLS moving forwards. If we need to, we could add additional options to individually select the rx and tx parts - or possibly a dedicated API function for it.

But is the "not that great" part a temporary issue that is being worked on? If we introduce such a feature we would have to support it for the long term - and in any case it won't be available until 3.1 at the earliest (which we haven't set a timeframe for yet).

**Comment by vvfedorenko** [comment]
> Well, we can't change it now for 3.0 and we at least need to continue to support SSL_OP_ENABLE_KTLS moving forwards. If we need to, we could add additional options to individually select the rx and tx parts - or possibly a dedicated API function for it.
> 
> But is the "not that great" part a temporary issue that is being worked on? If we introduce such a feature we would have to support it for the long term - and in any case it won't be available until 3.1 at the earliest (which we haven't set a timeframe for yet).

There is a work in progress to improve RX side of Linux Kernel TLS, but it greatly depends on kernel crypto API which doesn't fit properly for TLS. And even if we will get improvements reasonably fast it will never be backported, as all improvements are treated as new features. That's why all current users probably will not see it in near future. And I think that it's a good idea to provide users a way to choose which feature to use. I missed this PR during development cycle and thought that old `SSL_MODE_NO_KTLS_{RX,TX}` API was good enough, but probably should be switched to `SSL_MODE_KTLS_{RX,TX}` to make **default off** way of feature.

**Comment by mattcaswell** [comment]
I would prefer to wait to see if we really need this feature until we're further progressed with 3.1

**Pull request closed without merging**